### PR TITLE
[BUGFIX release] createRecord should only setup relationships it has …

### PR DESCRIPTION
**Pull request by stefanpenner**
…data for

relationships should in theory now only be accessed if they are used (`set` or `get`). This should mitigate the large number of test failures see during upgrade due to to https://github.com/emberjs/data/blob/e6cb564bdb33da4b3c15c3c668dc4b1fdafdf190/addon/-private/system/relationships/state/create.js#L20

If this doesn't actually fix the issue in enough cases, we can always turn that assert into a warn or something. But I would love a failing test for the exact case that remains (if one exists) to be 100% sure. As I would prefer to not loose a good assertion unless we must.

### CODE REVIEW

File: tests/integration/relationships/belongs-to-test.js [hunk 1]
```diff
@@ -970,26 +970,49 @@ test("belongsTo hasData sync not loaded", function(assert) {
...
 
-  run(function() {
-    var book = store.createRecord('book', { name: 'The Greatest Book' });
-    var relationship = book._internalModel._relationships.get('author');
+  run(() => {
+    let author = store.createRecord('author');
+    let book = store.createRecord('book', { name: 'The Greatest Book' });
+    let relationship = book._internalModel._relationships.get('author');
+
+    assert.equal(relationship.hasData, false, 'relationship has data');
```
Comment (hjdivad) [inline tests/integration/relationships/belongs-to-test.js]: assertion message is wrong: `relationship does not have data`

File: tests/integration/relationships/belongs-to-test.js [hunk 2]
```diff
@@ -970,26 +970,49 @@ test("belongsTo hasData sync not loaded", function(assert) {
...
-    var book = store.createRecord('book', { name: 'The Greatest Book' });
-    var relationship = book._internalModel._relationships.get('author');
+  run(() => {
+    let author = store.createRecord('author');
+    let book = store.createRecord('book', {
+      name: 'The Greatest Book'
+    });
+
+    let relationship = book._internalModel._relationships.get('author');
+    assert.equal(relationship.hasData, false, 'relationship has data');
```
Comment (hjdivad) [inline tests/integration/relationships/belongs-to-test.js]: see ^

File: tests/integration/relationships/has-many-test.js [hunk 3]
```diff
@@ -2547,25 +2547,44 @@ test("hasMany hasData sync not loaded", function(assert) {
...
 
-  run(function() {
-    var chapter = store.createRecord('chapter', { title: 'The Story Begins' });
-    var relationship = chapter._internalModel._relationships.get('pages');
+  run(() => {
+    let chapter = store.createRecord('chapter', { title: 'The Story Begins' });
+    let page = store.createRecord('page');
+
+    let relationship = chapter._internalModel._relationships.get('pages');
+    assert.equal(relationship.hasData, false, 'relationship has data');
```
Comment (hjdivad) [inline tests/integration/relationships/has-many-test.js]: see ^

File: tests/integration/relationships/has-many-test.js [hunk 4]
```diff
@@ -2547,25 +2547,44 @@ test("hasMany hasData sync not loaded", function(assert) {
...
 
 test("hasMany hasData sync created", function(assert) {
-  assert.expect(1);
+  assert.expect(2);
+
+  run(() => {
+    let chapter = store.createRecord('chapter', { title: 'The Story Begins' });
+    let relationship = chapter._internalModel._relationships.get('pages');
+
+    assert.equal(relationship.hasData, false, 'relationship has data');
```
Comment (hjdivad) [inline tests/integration/relationships/has-many-test.js]: see ^

**Comment by stefanpenner** [comment]
seems to be blocked by:

```js
SyntaxError: ember-data/adapters/json-api.js: @ember/debug does not have a deprecate import
   7 | import RESTAdapter from "./rest";
   8 | import { isEnabled } from '../-private';
>  9 | import { deprecate } from '@ember/debug';
     | ^
  10 | import { instrument } from 'ember-data/-debug';
```

@rwjblue is investigating.

**Comment by stefanpenner** [comment]
@runspired r?

**Comment by runspired** [comment]
LGTM

**Comment by stefanpenner** [comment]
@bmac got cycles for a backport + release? (I'm currently on vacation, but if you don't have cycles I can handle it)

**Comment by bmac** [comment]
@stefanpenner I'm on it.

**Comment by bmac** [comment]
Released as 2.14.4.

**Comment by stefanpenner** [comment]
@bmac awesome thanks dude, you rock!

**Comment by thomasdurin** [comment]
Hi,

Since this changes unset relationships are not serialized, but I was wondering if this match the JSON-API specs ?

Doesn't JSON-API specs always require to specify all relationships, including those that are null ?

**Pull request merged**